### ISSUE TITLE
[Feature Request] Add a Tick to activate an GPS Almanac Update over Internet

### ISSUE THREAD

**Issue by Mannshoch** [ubports/ubuntu-touch#834]
If you do not have an actual Almanac a GPS Fix could take long time. Also if you did not use GPS for a long time or change the location since last use. 
The GPS Almanac need around 12.5min for an update.

See also: https://stackoverflow.com/a/17876185/9730735

- Add a Tick for activating Almanac Update over Internet
- Add a possibility to setup an alternate server
- Maybe add a possibility to set the GPS into a "download Almanac only" mode. So GPS  itself is deactivated

**Comment by UniversalSuperBox** [comment]
The almanac can be obtained from https://www.navcen.uscg.gov/?pageName=gpsAlmanacs. This is a USA Government-run website. It has the files in tons of different formats, though. We'd need to know what format we need and where to put it on each device for this to be feasible.

**Comment by dobey** [comment]
I'm not sure what this is meant to solve? AFAIK, we do not need to bother with getting the almanac signal from a satellite (and well, first we must find and connect to a satellite), or getting it from alternate sources, in the code we need to concern ourselves with for UT.

If this is another complaint about the fact that we've removed the broken geoip, in the form of suggesting a "solution" with no context, I think we can just close it. Can you please describe a specific _problem_ instead? There is no such feature for downloading GPS almanac in Android either, AFAICS.

**Comment by Mannshoch** [comment]
Sorry I do not know the geoip subject and I also not know what you mean with the word UT. 
In my opinion the Problem is well described in this sentence:
> If you do not have an actual Almanac a GPS Fix could take long time. Also if you did not use GPS for a long time or change the location since last use.

I don't use ubuntu-touch since a while. One reason was that GPS as it had worked at the time, was not useful.

If you do not think that is a Problem please handle this ticket as Feature request. But please do not close it until that problem is solved. The Problem is solved If I'm able to start GPS and I get a position within 10sec even I didn't use it for a long time and I'm not near the last position I had GPS the last time.

**Comment by dobey** [comment]
> The Problem is solved If I'm able to start GPS and I get a position within 10sec even I didn't use it for a long time and I'm not near the last position I had GPS the last time.

This has nothing to do with a GPS almanac, and everything to do with complex AGPS solutions. You only get a fix so quickly on Android because it's using AGPS based on WiFi and Bluetooth devices around you. If you disable all of the extra AGPS features on Android, you will see too that it takes just as long to get a fix.

So, the problem you are trying to propose a solution for, is that GPS fix is not fast enough for your personal standards, under UT (ubuntu touch)?

**Comment by UniversalSuperBox** [comment]
Well, they're all about decreasing lock time in a way. This issue, as written, is about the almanac data. In a completely vanilla GPS system, it takes about 13 minutes to download [this data](https://en.wikipedia.org/wiki/GPS_signals#Navigation_message) from a satellite. That's why chips cache the data for a very long time. After it's downloaded once, it shouldn't need to be downloaded again (and it'll be re-cached the next time you use location for a while, I expect).

It's extremely difficult to tell how different vendors cache the almanac data, or even if we can upload an almanac to GPS hardware. It appears the NMEA command to do it is often proprietary, or it's a binary command that doesn't use any part of NMEA. It also appears that this process is part of a Qualcomm chipset's SOP, downloading the almanac from the carrier if it's connected to one (at least, [according to the fact that this system was insecure](https://www.theregister.co.uk/2016/12/07/android_qualcomm_move_on_insecure_gps_almanac_downloads/)). It's unclear what Mediatek or Exynos chipsets do.

https://github.com/ubports/ubuntu-touch/issues/1188 is about another way to decrease lock time, by using GPS location.

**Comment by dobey** [comment]
> Well, they're all about decreasing lock time in a way.

Certainly. But there's a huge difference between stating that lock time is slow ("it takes more than 30min to get a GPS lock" for example), versus asking for a feature that no other platform has, which one assumes will solve aforementioned slowness, based on some uninformed searching on the internet. The former act of actually describing problems is way better, as it lets us developers do the discovery work, rather than feature requests for things which may not actually solve the problem anyway.
